File: packages/otzaria_manager/lib/src/services/otzaria_release_client.dart
import 'dart:convert';
import 'dart:io' show Platform;

import 'package:http/http.dart' as http;
import 'package:otzaria_l10n/otzaria_l10n.dart';

import '../models/otzaria_release.dart';
import 'otzaria_asset_selector.dart';

/// שולף מידע על releases של github.com/Otzaria/otzaria (אפליקציית אוצריא
/// עצמה — לא SeforimLibrary/ה-DB).
class OtzariaReleaseClient {
  OtzariaReleaseClient({
    http.Client? httpClient,
    OtzariaTargetPlatform? platform,
    this.timeout = const Duration(seconds: 20),
  })  : _httpClient = httpClient ?? http.Client(),
        _platform =
            platform ?? OtzariaTargetPlatform.detect(Platform.operatingSystem);

  /// זמן קצוב לבקשה. חובה שיהיה כזה: בלעדיו, מחשב שמחובר לרשת אך בלי מסלול
  /// לאינטרנט (למשל captive portal) היה תולה את בדיקת העדכונים ללא הגבלה.
  /// ניתן לשינוי בזמן ריצה מהגדרות הלאנצ'ר.
  Duration timeout;

  // ⚠️ Otzaria/otzaria (ה-fork, לא Sivan22/otzaria המקורי) — זה הריפו
  // שממנו בפועל מפיצים releases ושהמשתמשים מורידים ממנו. תוקן אחרי
  // דיווח משתמש: הקוד דיווח 0.9.93 כשגרסה 0.9.95 כבר הייתה קיימת
  // ב-Otzaria/otzaria.
  static const _owner = 'Otzaria';
  static const _repo = 'otzaria';
  static const _apiBase = 'https://api.github.com';

  final http.Client _httpClient;

  /// פלטפורמת היעד שעבורה בוחרים אסט. ברירת המחדל היא הפלטפורמה שהלאנצ'ר
  /// רץ עליה; ניתן לדרוס אותה בבדיקות.
  final OtzariaTargetPlatform _platform;

  static const _assetSelector = OtzariaAssetSelector();

  /// כמה releases להביא כדי שיהיה מה לסנן. חיפוש ה-release היציב צריך
  /// לדלג על שרשרת ארוכה של preview builds לפני שהוא מגיע אליו.
  static const int _pageSize = 50;

  /// מחזיר את **שתי** הגרסאות שהלאנצ'ר מוריד: ה-release היציב האחרון
  /// (`prerelease=false`), ובנוסף ה-pre-release האחרון — אך ורק כשהוא חדש
  /// מהיציב. GitHub מחזיר את /releases מהחדש לישן, ולכן "חדש יותר" = מופיע
  /// לפניו ברשימה. draft נפסל תמיד.
  ///
  /// release שאין לו אסט התקנה לפלטפורמה הנוכחית מדולג במקום להפיל את כל
  /// הבדיקה — הערוץ ממשיך לגרסה הוותיקה יותר. אם אף גרסה לא ניתנת להתקנה,
  /// נזרקת [NoInstallerAssetException] של החדשה ביותר שנפסלה.
  Future<OtzariaChannelReleases> fetchChannelReleases() async {
    final candidates = await _fetchReleasesJson();

    OtzariaRelease? stable;
    OtzariaRelease? prerelease;
    NoInstallerAssetException? firstRejection;

    for (final json in candidates) {
      final isPrerelease = json['prerelease'] as bool? ?? false;
      // כבר יש pre-release חדש יותר; ותיקים ממנו אינם מעניינים.
      if (isPrerelease && prerelease != null) continue;

      final OtzariaRelease release;
      try {
        release = _parseRelease(json);
      } on NoInstallerAssetException catch (e) {
        firstRejection ??= e;
        continue;
      }

      if (!isPrerelease) {
        // הגענו ליציב: כל מה שמתחתיו ברשימה ותיק ממנו.
        stable = release;
        break;
      }
      prerelease = release;
    }

    if (stable == null && prerelease == null && firstRejection != null) {
      throw firstRejection;
    }
    return OtzariaChannelReleases(stable: stable, prerelease: prerelease);
  }

  /// ה-releases הגולמיים, מהחדש לישן, בלי draft.
  Future<List<Map<String, dynamic>>> _fetchReleasesJson() async {
    final uri = Uri.parse(
      '$_apiBase/repos/$_owner/$_repo/releases?per_page=$_pageSize',
    );
    final response = await _httpClient.get(
      uri,
      headers: const {
        'Accept': 'application/vnd.github+json',
        'X-GitHub-Api-Version': '2022-11-28',
        // ⚠️ חובה: GitHub API מחזיר 403 "forbidden by administrative
        // rules" לכל בקשה בלי User-Agent — ללא קשר ל-rate limit.
        'User-Agent': 'otzaria-launcher',
      },
    ).timeout(timeout);

    if (response.statusCode != 200) {
      throw GithubApiException(
        AppL10n.strings.appDomain.githubStatus(response.statusCode, '$uri'),
      );
    }

    final decoded = jsonDecode(response.body) as List<dynamic>;
    if (decoded.isEmpty) {
      throw StateError(
        AppL10n.strings.appDomain.noReleasesAtAll('$_owner/$_repo'),
      );
    }

    return decoded
        .cast<Map<String, dynamic>>()
        .where((r) => !(r['draft'] as bool? ?? false))
        .toList(growable: false);
  }

  OtzariaRelease _parseRelease(Map<String, dynamic> json) {
    final tagName = json['tag_name'] as String;
    final assets =
        (json['assets'] as List<dynamic>).cast<Map<String, dynamic>>();

    // בחירת האסט לפי פלטפורמת היעד — ראו [OtzariaAssetSelector] להסבר על
    // כללי ההתאמה (ולמה חבילות ה-FULL של 2GB נפסלות מעצמן).
    final selected = _assetSelector.select(
      platform: _platform,
      assets: assets,
      nameOf: (asset) => asset['name'] as String,
    );

    if (selected == null) {
      throw NoInstallerAssetException(
        tagName: tagName,
        platform: _platform,
        expectedSuffixes: OtzariaAssetSelector.expectedSuffixesFor(_platform),
      );
    }

    final (asset, installerKind) = selected;

    // חבילת ה-FULL נקראת תמיד מהמטא־דאטה, גם כשלא מורידים אותה: בלי זה
    // הגדרה שנדלקת הייתה דורשת בדיקה חוזרת מול GitHub כדי לגלות שהיא בכלל
    // קיימת. ההורדה עצמה מותנית בהגדרה — ראו `OtzariaAppMirror.sync`.
    final fullAsset = _assetSelector.selectFull(
      platform: _platform,
      assets: assets,
      nameOf: (asset) => asset['name'] as String,
    );

    return OtzariaRelease(
      tagName: tagName,
      name: (json['name'] as String?) ?? tagName,
      isPrerelease: json['prerelease'] as bool? ?? false,
      isDraft: json['draft'] as bool? ?? false,
      publishedAt: json['published_at'] == null
          ? null
          : DateTime.tryParse(json['published_at'] as String),
      installerKind: installerKind,
      installerAssetName: asset['name'] as String,
      installerDownloadUrl: asset['browser_download_url'] as String,
      installerSizeBytes: asset['size'] as int,
      fullPackage: fullAsset == null
          ? null
          : OtzariaFullPackage(
              assetName: fullAsset.$1['name'] as String,
              downloadUrl: fullAsset.$1['browser_download_url'] as String,
              sizeBytes: fullAsset.$1['size'] as int,
              installerKind: fullAsset.$2,
            ),
      releaseNotes: json['body'] as String?,
    );
  }

  void dispose() => _httpClient.close();
}

/// שגיאת תגובה לא תקינה מ-GitHub API. שם ייעודי (לא HttpException) כדי לא
/// להתנגש עם dart:io.HttpException אם הצרכן מייבא גם את dart:io ישירות.
class GithubApiException implements Exception {
  const GithubApiException(this.message);
  final String message;

  @override
  String toString() => message;
}

File: native/web/js/store/otzaria_release_client.js
// מברר מהן גרסאות **אוצריא** שהמראה נבנית עבורן, מתוך ה-releases של
// `Otzaria/otzaria`.
//
// ── למה בכלל, ולא הגרסה שבמחשב ───────────────────────────────────────────────
// המראה נבנית במחשב מקוון ונצרכת במחשב **אחר**, לא-מקוון. סינון לפי
// הגרסה שבמחשב המסנכרן פירושו שכונן שנבנה על מחשב עם אוצריא ישנה נושא
// בילדים ישנים, ומחשב היעד — שגרסתו אחרת — מקבל "טרם ירד, יש לבצע
// סנכרון" בדיוק במקום שאין בו אינטרנט. לכן היעד הוא מה שהריפו מפרסם,
// ולא מה שמותקן כאן.
//
// ── שתי גרסאות, ולא אחת ──────────────────────────────────────────────────────
// כשה-release האחרון מסומן `prerelease`, מי שעל ערוץ היציב אינו מריץ
// אותו — ובילד שנבחר לפיו לא בהכרח יעלה אצלו. לכן במצב הזה מתבררות
// **שתיהן**: האחרונה, והיציבה האחרונה. `resolveTargets` יגזור מהן בילד
// לכל אחת, וכשהשתיים נפתרות לאותו בילד יורד אחד.
//
// ⚠️ **הדגל, לא השם.** בריפו הזה כל ה-releases נקראים
// "Otzaria X (Preview from dev)" — כולל אלה שאינם prerelease. השם אינו
// סימן לכלום, והדגל `prerelease` הוא הנתון היחיד שאפשר להכריע לפיו.

import {comparePluginVersions} from '../util/version.js';

const OWNER = 'Otzaria';
const REPO = 'otzaria';
const API_BASE = 'https://api.github.com';

/**
 * דף אחד. הגרסה האחרונה והיציבה האחרונה נמצאות בראש הרשימה בפועל, ודף
 * שני היה מאריך את הסנכרון בשביל היסטוריה שאינה מעניינת אותנו.
 */
const PAGE_SIZE = 30;

/**
 * תג גרסה של אוצריא: `0.9.96+736`, `v0.9.96`, `0.9.96`.
 *
 * ⚠️ מקטע prerelease (`-pr-715`, `-dev-110`) **נפסל**: אלה בילדים של
 * בקשות משיכה ושל dev, ולא גרסה שמישהו מריץ. הם אמורים לשאת גם את הדגל
 * `prerelease`, אבל בריפו הזה יש בפועל תגים כאלה בלי הדגל — ולכן התג
 * נבדק בנפרד ולא נסמך עליו.
 */
const RELEASE_TAG = /^v?(\d+(?:\.\d+)*)(?:\+\d+)?$/;

/** דף ה-releases, כשצריך להצביע על מקום ולא על גרסה. */
export const OTZARIA_RELEASES_PAGE_URL =
    `https://github.com/${OWNER}/${REPO}/releases`;

/**
 * הגרסה שבתוך התג, בלי `v` ובלי מספר הבילד (`+736`), או `null` כשאינו
 * תג גרסה.
 *
 * מספר הבילד נחתך כי הוא **אינו** חלק מהגרסה שהתוספים מצהירים עליה:
 * `compatibleWith` באתר הוא `0.9.96`, וההשוואה מולו צריכה להיות מול
 * אותו מספר בדיוק.
 */
export function otzariaVersionOf(tag) {
  const match = RELEASE_TAG.exec(String(tag ?? '').trim());
  return match === null ? null : match[1];
}

export class OtzariaReleaseClient {
  constructor({net, timeoutMs = 20000}) {
    this.net = net;
    /**
     * זמן קצוב — חובה: בלעדיו מחשב שמחובר לרשת בלי מסלול לאינטרנט היה
     * תולה את הסנכרון כולו לפני שהוא מתחיל.
     */
    this.timeoutMs = timeoutMs;
  }

  /**
   * הגרסאות שהמראה תיבנה עבורן, מהגבוהה לנמוכה: הגרסה האחרונה, ואם היא
   * `prerelease` — גם היציבה האחרונה שמתחתיה.
   *
   * רשימה **ריקה** כשאין אף release עם תג גרסה. המתקשר מתייחס אליה
   * כ"אין מול מה לסנן", בדיוק כמו לכשל.
   *
   * @returns {Promise<{versions: string[], latest: string|null,
   *                    latestStable: string|null,
   *                    latestIsPrerelease: boolean}>}
   */
  async fetchTargetVersions() {
    const url =
        `${API_BASE}/repos/${OWNER}/${REPO}/releases?per_page=${PAGE_SIZE}`;
    // ⚠️ GitHub מחזיר 403 לכל בקשה בלי User-Agent. ה-host קובע אותו
    // ברמת הסשן (`kUserAgent` ב-netapi.cpp), ולכן אין צורך בכותרת כאן.
    const response = await this.net.get(url, this.timeoutMs);
    if (response.status !== 200) {
      throw new Error(`GitHub החזיר ${response.status} עבור ${url}`);
    }

    const decoded = JSON.parse(response.body);
    if (!Array.isArray(decoded)) {
      throw new Error(`אין releases ב-${OWNER}/${REPO}`);
    }
    return pickTargetVersions(decoded);
  }
}

/**
 * הבחירה עצמה, בלי רשת — חשוף כדי שבדיקות יאמתו אותה על תשובה מוקלטת.
 *
 * הגבוהה ולא הראשונה ברשימה: סדר הפרסום אינו סדר הגרסאות (release
 * שנערך ידנית, תיקון שפורסם לגרסה ותיקה אחרי החדשה), ובחירת "הראשון"
 * הייתה תלוית-מזל.
 *
 * @param {Array<object>} releases תשובת `/releases` של GitHub
 */
export function pickTargetVersions(releases) {
  let latest = null;
  let latestStable = null;

  for (const entry of releases) {
    if (entry === null || typeof entry !== 'object') continue;
    // draft אינו קיים לציבור; תג שאינו גרסה (או בילד PR/dev) אינו גרסה
    // שמישהו מריץ.
    if (entry.draft === true) continue;
    const version = otzariaVersionOf(entry.tag_name);
    if (version === null) continue;

    if (latest === null || comparePluginVersions(version, latest) > 0) {
      latest = version;
    }
    if (entry.prerelease !== true &&
        (latestStable === null ||
         comparePluginVersions(version, latestStable) > 0)) {
      latestStable = version;
    }
  }

  const latestIsPrerelease = latest !== null &&
      (latestStable === null || comparePluginVersions(latest, latestStable) > 0);

  // סדר יורד, כדי שהצרכנים יקבלו את הגרסאות במפורש מהגבוהה לנמוכה.
  const versions = [];
  if (latest !== null) versions.push(latest);
  if (latestIsPrerelease && latestStable !== null) {
    versions.push(latestStable);
  }

  return {versions, latest, latestStable, latestIsPrerelease};
}
